real_estate/models/estate_property.py

```diff
@@ -0,0 +1,211 @@
+        string="Salesperson",
+        default=lambda self: self.env.user
+    )
+
+    tag_ids = fields.Many2many(
+        comodel_name='estate.property.tag',
+        string="Tag"
+    )
+
+    offer_ids = fields.One2many(
+        comodel_name='estate.property.offer',
+        inverse_name='property_id',
+        string="Offers"
+    )
+
+    total_area = fields.Float(
+        string="Total Area",
+        compute='_compute_total_area'
+    )
+
+    @api.depends('garden_area', 'living_area')
+    def _compute_total_area(self):
+        for property in self:
+            property.total_area = property.garden_area + property.living_area
+
+    best_price = fields.Float(
+        string="Best Offer",
+        compute='_compute_best_price',
+    )
+
+    @api.depends('offer_ids.price')
+    def _compute_best_price(self):
+        for property in self:
+            if property.offer_ids:
+                property.best_price = max(property.offer_ids.mapped('price'))
+            else:
+                property.best_price = 0.0
+
+    @api.onchange('garden')
+    def _onchange_garden(self):
+        if self.garden:
+            self.garden_orientation = 'north'
+            self.garden_area=10
+        else:
+            self.garden_orientation = False
+            self.garden_area = 0
+
+    def action_set_sold(self):
+        for property in self:
+            if property.state == "cancelled":
+                raise UserError("Cancelled Property can not be sold")
+            property.state = 'sold'
+        
+    def action_set_cancel(self):
+        for property in self:
+            if property.state == "sold":
+                raise UserError("Sold Property can not be cancel")
+            property.state = 'cancelled'
+
+    # SQL CONSTRAINTS
+    _sql_constraints = [
+        (
+            'check_expected_price',
+            'CHECK(expected_price>0)',
+            'Expected price must be strictly positive'
+        ),
+        (
+            'check_selling_price',
+            'CHECK(selling_price>0)',
+            'Selling price must be positive'
+        )
+    ]
+
+    # PYTHON CONSTRAINTS
+    @api.constrains('expected_price', 'selling_price')
+    def _check_selling_price(self):
+        for product in self:
+            min_selleing_price = 0.9 * product.expected_price
+            if (not float_is_zero(product.selling_price, precision_digits=2) and 
+                float_compare(product.selling_price, min_selleing_price, 
+                precision_rounding=2) == -1
+                ):
+                raise ValidationError("Selling Price cannot be lower than 90% of the Expected Price.")
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_check_property_state(self):
+        for property in self:
+            if property.state not in ['new', 'cancelles']:
+                d = {
+                    'offer_received':'Offer Received',
+                    'offer_accepted':'Offer Accepted',
+                }
```

Comment:
```suggestion
                state_mapping = {
                    k: v
                    for k, v self._fields['state']._description_selection(self.env)
                }
```
bad naming convention and can be fetched from fields, `dict` better be defined before the loop starts

Reply:
Sorry sir! I was just checking to see the if the code provide expected output  or not and then forgot to change